### PR TITLE
drivers: clock_control_nrf: Fix releasing/stopping of HFCLK

### PR DESCRIPTION
**Pull request by anangl**
After switching to the new ZTEST API, the `test_bt_interrupted` test in the `nrf_onoff_and_bt` suite for the clock_control driver started failing. This revealed a bug in this test (see the "tests: clock_control: nrf_onoff_and_bt: Add clearing of `test_end`" commit for details) and after it was fixed, it turned out that the HFCLK stopping routines in the clock_control_nrf driver do not entirely work as they are supposed to (and as that broken test falsely confirmed).
This PR fixes both the test (with some additional clean-up) and the driver.

### CODE REVIEW

**Comment by anangl** [comment]
@cvinayak Could you please check if these changes do not significantly degrade the BLE controller performance as it was in #24334.

**Review by nordic-krch — APPROVED** [review]

File: drivers/clock_control/clock_control_nrf.c [hunk 1]
```diff
@@ -295,12 +295,21 @@ static void generic_hfclk_start(void)
...
+	 * before hfclk_stop() is called and the preempting code could request
+	 * the HFCLK again. Then, the HFCLK would be stopped inappropriately
+	 * and hfclk_user would be left with an incorrect value.
+	 */
+	unsigned int key = irq_lock();
+
+	hfclk_users &= ~HF_USER_GENERIC;
+	/* Skip stopping if BT is still requesting the clock. */
+	if (!(hfclk_users & HF_USER_BT)) {
+		hfclk_stop();
```
Comment (cvinayak) [inline drivers/clock_control/clock_control_nrf.c:309]: I have not tested on target with ISR CPU time profiling but I am concerned with `nrfx_clock_hfclk_stop` that has `while` loop without putting CPU in idle.

Any reason you can think that prevents the clock from being stopped in a asynchronous way?

Comment (anangl) [reply drivers/clock_control/clock_control_nrf.c:309]: Well, currently there is no other option of stopping the clock using the nrfx driver. I'm not sure about the reasons and if an asynchronous way could be added. Perhaps @nika-nordic or @nordic-krch could say something more in this regard. But anyway, this is actually beyond the scope of this PR.

Comment (cvinayak) [reply drivers/clock_control/clock_control_nrf.c:309]: > this is actually beyond the scope of this PR.

Well the change in this PR is enclosing a `while` loop inside `irq_lock()` and `irq_unlock()` which seems a bit dangerous in terms of OS being prevented from covering if anything goes wrong.

Comment (nika-nordic) [reply drivers/clock_control/clock_control_nrf.c:309]: There are no peripheral events associated with the clock being stopped so asynchronous clock stopping cannot be done in `nrfx_clock`

Comment (nordic-krch) [reply drivers/clock_control/clock_control_nrf.c:309]: Isn't it that we can stuck in this loop in two case: loop is preempted and someone starts the clock after preemption or hardware failure. Lock prevents first scenario and regarding the second one, we are screwed anyway.

**Comment by cvinayak** [comment]
PR:
image

main 4c70bf6630:
image

**Review by cvinayak — APPROVED** [review]

**Comment by nordic-piks** [comment]
@anangl Does this fix #49996?

**Comment by anangl** [comment]
> @anangl Is this fix for #49996?

Yes, for the nrf_onoff_and_bt part of the issue. And the nrf_lf_clock_start* tests should be fixed by #52333.

**Pull request merged**